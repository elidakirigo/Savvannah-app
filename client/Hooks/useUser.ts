'use client'

import axios from 'axios'
import { useEffect, useState } from 'react'
import { Bounce, ToastContainer, toast } from 'react-toastify' 

/**
 * Fetch user data from api returning all users
 * @returns an array of all users
 */

export const useUser = () => {
	const [user, setUser] = useState([])

	useEffect(() => {
		
		const fetchUser = async () => {
			const { data: users } = await toast.promise(
				axios.get(process.env.NEXT_PUBLIC_URL + '/users'),
				{
					pending: 'loading users ',
					success: 'user loaded successfully',
					error: 'No user found ',
				},
				{
					position: 'top-center',
					autoClose: 3000,
					hideProgressBar: false,
					closeOnClick: true,
					pauseOnHover: true,
					draggable: true,
					progress: undefined,
					theme: 'light',
					transition: Bounce,
				},
			)
			setUser(users)
		}

		fetchUser()
	}, [])

	return user
}
