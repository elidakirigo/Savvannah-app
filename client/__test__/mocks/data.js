export const users = [{ address: { street: 'Kulas Light', suite: 'Apt. 556', city: 'Gwenborough', zipcode: '92998-3874' }, company: { name: 'Romaguera-Crona', catchPhrase: 'Multi-layered client-server neural-net', bs: 'harness real-time e-markets' }, email: 'Sincere@april.biz', id: 1, name: 'Leanne Graham', phone: '1-770-736-8031 x56442', username: 'Bret', website: 'hildegard.org' }]

export const mockSession = {
	expires: new Date(Date.now() + 2 * 86400).toISOString(),
	user: { name: 'moon pie', email: 'wakamoonpie@gmail.com', image: 'https://lh3.googleusercontent.com/a/ACg8ocLTSwAi2PRqfpYSuak05Q2jiRKBnYqL4wv88UJqLxzmel3jXw=s96-c' },
}

export const mockUsers = [
	{
		id: 1,
		name: 'Leanne Graham',
		username: 'Bret',
		email: 'Sincere@april.biz',
		address: {
			street: 'Kulas Light',
			suite: 'Apt. 556',
			city: 'Gwenborough',
			zipcode: '92998-3874',
			geo: {
				lat: '-37.3159',
				lng: '81.1496',
			},
		},
		phone: '1-770-736-8031 x56442',
		website: 'hildegard.org',
		company: {
			name: 'Romaguera-Crona',
			catchPhrase: 'Multi-layered client-server neural-net',
			bs: 'harness real-time e-markets',
		},
	},
]

export const mocksetUsers = jest.fn()

export const allUserdata = [
	{
		currentAlbums: [
			{ userId: 1, id: 1, title: 'quidem molestiae enim' },
			{ userId: 1, id: 2, title: 'sunt qui excepturi placeat culpa' },
			{ userId: 1, id: 3, title: 'omnis laborum odio' },
			{ userId: 1, id: 4, title: 'non esse culpa molestiae omnis sed optio' },
			{ userId: 1, id: 5, title: 'eaque aut omnis a' },
			{ userId: 1, id: 6, title: 'natus impedit quibusdam illo est' },
			{ userId: 1, id: 7, title: 'quibusdam autem aliquid et et quia' },
			{ userId: 1, id: 8, title: 'qui fuga est a eum' },
			{ userId: 1, id: 9, title: 'saepe unde necessitatibus rem' },
			{ userId: 1, id: 10, title: 'distinctio laborum qui' },
		],
		email: 'Sincere@april.biz',
		id: 1,
		name: 'Leanne Graham',
		noOfAlbums: 10,
		phone: '1-770-736-8031 x56442',
		username: 'Bret',
		website: 'hildegard.org',
	},
]

export const AlbumPhotos = [[{ albumId: 1, id: 1, thumbnailUrl: 'https://via.placeholder.com/150/92c952', title: 'accusamus beatae ad facilis cum similique qui sunt', url: 'https://via.placeholder.com/600/92c952' }]]

export const AllPhotos = [{ albumId: 1, id: 1, thumbnailUrl: 'https://via.placeholder.com/150/92c952', title: 'accusamus beatae ad facilis cum similique qui sunt', url: 'https://via.placeholder.com/600/92c952' }]

export const UserData = [{ users: allUserdata[0] }]

export const user = allUserdata

module.exports = { mockSession, allUserdata, AlbumPhotos, AllPhotos, UserData, user }
