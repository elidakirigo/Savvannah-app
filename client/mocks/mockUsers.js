const users = [{ address: { street: 'Kulas Light', suite: 'Apt. 556', city: 'Gwenborough', zipcode: '92998-3874' }, company: { name: 'Romaguera-Crona', catchPhrase: 'Multi-layered client-server neural-net', bs: 'harness real-time e-markets' }, email: 'Sincere@april.biz', id: 1, name: 'Leanne Graham', phone: '1-770-736-8031 x56442', username: 'Bret', website: 'hildegard.org' }]

export default async function mockFetch(url) {
	if (url.startsWith('https://api.nationalize.io') && url.includes('john')) {
		return {
			ok: true,
			status: 200,
			json: async () => nameNationalizeResponse,
		}
	}

	throw new Error(`Unhandled request: ${url}`)
}
