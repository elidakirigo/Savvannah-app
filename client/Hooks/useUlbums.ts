'use client'

import axios from 'axios'
import { useEffect, useState } from 'react'
import { useUser } from './useUser'

export const useAlbums = () => {
	const [album, setAlbum] = useState([])

	useEffect(() => {
		const fetchAlbums = async () => {
			const { data: albums } = await axios.get(process.env.NEXT_PUBLIC_URL + '/albums')
			setAlbum(albums)
		}

		fetchAlbums()
	}, [])

	return album
}

export const useUserAlbum = () => {
	const albums = useAlbums()
	const user = useUser()

	const userIndex = user.map((user: { id: number }) => user.id)

	const UserAlbums = userIndex.map((data) => {
		return albums.filter((a: { userId: number }) => {
			if (data == a.userId) {
				return { data: { ...a } }
			}
		})
	})

	const allUserdata = user.map((data: { id: number; email: string; name: string }) => {
		const noOfAlbums = UserAlbums[data.id - 1]?.length
		const currentAlbums = UserAlbums[data.id - 1]
		return { ...data, noOfAlbums, currentAlbums }
	})

	return { UserAlbums, allUserdata }
}
