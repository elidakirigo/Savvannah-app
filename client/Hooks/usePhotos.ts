'use client'

import axios from 'axios'
import { useEffect, useState } from 'react'
import { useAlbums } from './useUlbums'
import { useAppSelector } from '@/app/store/hooks'
import { Bounce, toast } from 'react-toastify'
import { useDispatch } from 'react-redux'
import { fetchphotos } from '@/app/store/slices/photoSlice'
/**
 * Fetches the photos from the api
 * @returns an array of all photos
 */

export const useFetchPhotos = () => {
	const [photos, setPhotos] = useState([])
	const dispatch = useDispatch()

	useEffect(() => {
		const fetchphoto = async () => {
			const { data: photos } = await toast.promise(
				axios.get(process.env.NEXT_PUBLIC_URL + '/photos'),
				{
					pending: 'loading photos ',
					success: 'photos loaded successfully',
					error: 'No photos found ',
				},
				{
					position: 'top-center',
					autoClose: 3000,
					hideProgressBar: false,
					closeOnClick: true,
					pauseOnHover: true,
					draggable: true,
					progress: undefined,
					theme: 'light',
					transition: Bounce,
				},
			)
			setPhotos(photos)
			dispatch(fetchphotos(photos))

		}

		fetchphoto()
	}, [])

	return photos
}

/**
 * fetches data from redux store and sends to home
 * @returns  photos that were fetched from redux
 */

export const usePhotos = () => {
	const photoData = useAppSelector((state) => state.photos)
	const photo = useFetchPhotos()

	if (photoData.photos) return photoData.photos

	return photo
}
/**
 * Organizes the photos according to same albums
 * @returns an array of organized objects with photos that share the same album
 */
export const useAlbumPhotos = () => {
	const album = useAlbums()
	const photos = usePhotos()
	const AlbumIndex = album.map((album: { id: number }) => album.id)

	const Albums = AlbumIndex.map((data) => {
		return photos.filter((albumdata: { albumId: number }) => {
			if (data == albumdata.albumId) {
				return { data: { ...albumdata } }
			}
		})
	})

	return Albums
}
