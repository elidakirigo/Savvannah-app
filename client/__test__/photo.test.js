import '@testing-library/jest-dom'
import { createMemoryHistory } from 'history'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import * as nextAuthReact from 'next-auth/react'
import { useState } from 'react'
import Link from 'next/link'
import Photo from '../src/app/photo/[id]/photo'
import StoreProvider from '../components/StoreProvider'
import { mockSession, AllPhotos } from './mocks/data'
import { usePhotos } from '../Hooks/usePhotos'

jest.mock('next-auth/react')

jest.mock('../Hooks/usePhotos')

jest.mock('next/navigation', () => {
	return {
		__esModule: true,
		usePathname: () => ({
			pathname: '',
		}),
		useRouter: () => ({
			push: jest.fn(),
			replace: jest.fn(),
			prefetch: jest.fn(),
		}),
		useSearchParams: () => ({
			get: () => {},
		}),
	}
})

// jest.mock('react', () => ({
// 	...jest.requireActual('react'),
// 	useState: jest.fn(),
// }))

const nextAuthReactMocked = nextAuthReact

// const useStateMock = useState

describe('renders the photo page', () => {
	// const setTitleNametitleName = jest.fn()

	// useStateMock.mockImplementation(() => ['titleName', setTitleNametitleName])

	beforeEach(() => {
		nextAuthReactMocked.useSession.mockImplementation((_options) => {
			return { data: mockSession, status: 'authenticated' }
		})

		nextAuthReactMocked.signIn.mockImplementation(() => Promise.resolve({ error: '', status: 200, ok: true, url: '' }))

		usePhotos.mockReturnValue(AllPhotos)

		render(
			<StoreProvider>
				<Photo currentId='1' />
			</StoreProvider>,
		)
	})

	// afterEach(() => {
	// 	jest.restoreAllMocks()
	// })

	it('should render Albumpage successfully', async () => {
		await waitFor(() => {
			const { container } = render(
				<StoreProvider>
					<Photo currentId='1' />
				</StoreProvider>,
			)
			expect(container).toMatchSnapshot()
		})
	})

	it('should have Image ', async () => {
		await waitFor(() => {
			const image = screen.getByRole('img')
			expect(image).toBeInTheDocument()
		})
	})

	// it('should have image title', async () => {
	// 	await waitFor(() => {
	// 		expect(screen.getAllByTestId('image-title')[0]).toHaveTextContent('accusamus')
	// 	})
	// })

	it('should have image src', async () => {
		await waitFor(() => {
			const testImage = document.querySelector('img')
			expect(testImage.src).toContain('https://')
		})
	})
})
