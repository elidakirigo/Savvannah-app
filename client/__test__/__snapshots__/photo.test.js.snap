// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the photo page should render Albumpage successfully 1`] = `
<div>
  <div
    class="flex p-5 items-center gap-4 flex-col"
  >
    <div
      class="max-w-[400px]"
    >
      <img
        alt="accusamus beatae ad facilis cum similique qui sunt"
        class="w-full"
        data-nimg="1"
        data-testid="image"
        decoding="async"
        height="10"
        loading="lazy"
        src="https://via.placeholder.com/150/92c952"
        style="color: transparent;"
        width="10"
      />
    </div>
    <h1
      data-testid="image-title"
    >
      accusamus beatae ad facilis cum similique qui sunt
    </h1>
    <button
      class="bg-purple-700 py-2 text-white cursor-pointer hover:bg-purple-300 hover:border-2 hover:border-purple-900 hover:text-purple-900 px-6 rounded-lg shadow-md"
      data-testid="editbtn"
    >
      edit title
    </button>
  </div>
</div>
`;
